Add awslogs driver for Amazon CloudWatch Logs

samuelkarp opened the pull request:
This pull request adds a new logging driver `awslogs` which sends log events to Amazon CloudWatch Logs.

samuelkarp committed:
Vendor AWS Go SDK

Signed-off-by: Samuel Karp <skarp@amazon.com>

samuelkarp committed:
Add awslogs driver for Amazon CloudWatch Logs

Signed-off-by: Samuel Karp <skarp@amazon.com>

calavera commented:
Thanks for adding this @samuelkarp, I :heart: the amount of tests in this PR. We are trying to document each method we add, even unexported ones, do you mind to add the missing docs?

/cc @LK4D4

samuelkarp commented:
@calavera I've added comments to each method.

calavera commented:
I think we also need to add documentation to the log-driver flags to add this new driver.

samuelkarp commented:
@calavera I've updated the pull request according to your comments.

calavera commented:
LGTM.

/cc @LK4D4

cpuguy83 commented:
Can you update to make sure all names are camelized?
For instance `cwlogstream` should be `cwLogStream`, and probably more appropriately `logStream`
`cwlogsiface` should be `logger`, etc.

calavera commented:
If you add your new package to the linter you'll see right away what's missing:

https://github.com/docker/docker/blob/master/hack/make/validate-lint#L34

LK4D4 commented:
I agree with @cpuguy83

samuelkarp commented:
I'm seeing some Jenkins failures that appear to be unrelated to my changes, but I believe I've addressed the comments here.

@LK4D4, I've addressed all of your [concerns](https://github.com/samuelkarp/docker/commit/3c365b32ff0dae6ffb7597af2f296f4709e35aae) except for leaving `logStreamName` as-is to match the service.

@cpuguy83, I've renamed `cwlogstream` to `logStream` and `cwlogsiface` to `api` (I don't want to shadow the `logger` package and it's really just the API for the service).

@calavera, I had added my new package to the linter before and it appears to pass without complaint.

cpuguy83 commented on daemon/logger/awslogs/cloudwatchlogs.go:
```diff
@@ -0,0 +1,330 @@
+type logStream struct {
+	logStreamName string
+	logGroupName  string
+	cwlogsclient  api
```
maybe just `client`?

cpuguy83 commented on daemon/logger/awslogs/cloudwatchlogs.go:
```diff
@@ -0,0 +1,330 @@
+	PutLogEvents(*cloudwatchlogs.PutLogEventsInput) (*cloudwatchlogs.PutLogEventsOutput, error)
+}
+
+type sortableevents []*cloudwatchlogs.InputLogEvent
```
maybe just `events`

samuelkarp replied:
Sure, though I originally had this as `sortableevents` as the only reason I defined the type is to implement the `sort.Interface` interface.

euank replied:
One naming idea I've seen here (first from Dave Cheney) is to name it `byTimestamp` so that you later write `sort.Sort(byTimestamp(events))`. Another idea for naming would be `logEventSlice` to follow the `*Slice` types in the std lib sort package.

LK4D4 replied:
I agree with @euank

cpuguy83 commented on daemon/logger/awslogs/cloudwatchlogs.go:
```diff
@@ -0,0 +1,330 @@
+			}
+			if awsErr.Code() == resourceAlreadyExistsCode {
+				// Allow creation to succeed
+				logrus.WithFields(fields).Warn("Log stream already exists")
```
Debug level may be better here.

samuelkarp replied:
I think I'd lean more toward Info level, as multiple containers writing to the same log stream at the same time can have performance implications.

cpuguy83 commented on daemon/logger/awslogs/cloudwatchlogs.go:
```diff
@@ -0,0 +1,330 @@
+
+	_, err := l.cwlogsclient.CreateLogStream(input)
+
+	if err != nil {
```
We probably don't need to log this error since it can/should be handled by the caller.

samuelkarp commented:
@cpuguy83 I've addressed your comments, except that I've left `sortableevents` as-is to avoid conflict with the `events` variable.

samuelkarp commented:
@euank thanks for the suggestion! @cpuguy83 @LK4D4 I've renamed `sortableevents` to `byTimestamp` so the sort looks like `sort.Sort(byTimestamp(events))`.

cpuguy83 commented:
@samuelkarp Oh that's actually quite fantastic!

LK4D4 commented:
LGTM
Pls test this with aws someone.

LK4D4 commented on docs/reference/logging/awslogs.md:
```diff
@@ -0,0 +1,89 @@
+To configure the region to be used, specify either the `aws-region` log option
+or `AWS_REGION` environment variable:
+
+    docker run --log-driver=awslogs --log-opt aws-region=us-east-1 ...
```
I have concern about opt names. I think maybe we can rename `log-driver` to `aws` or prepend all logs with `awslogs` for consistency.

samuelkarp commented:
@LK4D4 I've changed the option to `awslogs-region`.

samuelkarp commented:
Hi @calavera, @cpuguy83, and @LK4D4, what are the next steps here?

LK4D4 commented:
@samuelkarp Plan was that someone from maintainer could test it against aws :)

samuelkarp commented:
@LK4D4, please let me know if there’s anything I can do to help facilitate your testing.

calavera commented:
I just tested this and it seems to work as expected.

This is the cloudwatch console with my logs:

![image](https://cloud.githubusercontent.com/assets/1050/9599352/7e808628-5093-11e5-98d0-a8383c88d788.png)

There is only one thing worth mentioning, the AWS credentials must be set for the daemon, not in the client. I think we should mention that in the documentation because people might be confused with this. Even if you start a container with `docker run --log-driver=awslogs`, those credentials must be set for the daemon.

calavera commented:
@cpuguy83, @LK4D4 feel free to move this to docs review if it still looks good to you.

@samuelkarp It would be great if you could add an extra line in the docs about the credentials for the daemon in preparation for the docs review.

samuelkarp commented:
@calavera Thanks for testing! I've updated the language around credentials to try and make it more clear that credentials must be supplied to the daemon.  I'm seeing some Jenkins failures again that appear to be unrelated to my changes.

calavera commented:
Thanks, it still needs one more review from a maintaner though.

LK4D4 commented:
ok LGTM
ping @thaJeztah @moxiegirl

thaJeztah commented:
Added some minor comments on the docs, thanks @samuelkarp !

samuelkarp commented:
@thaJeztah Thanks for taking a look!  I've updated the docs to address your comments.

thaJeztah commented:
Thanks @samuelkarp, I tested the docs (`make docs`), and it looks good.

LGTM

calavera commented:
@samuelkarp this is going to need a rebase :pensive:

samuelkarp commented:
@calavera, I've rebased.  I'm expecting Jenkins to fail as 6990b76a696dd265674f4c2973f25755a6485f05 changed `CDmTask` to `cdmTask` but didn't update `/pkg/devicemapper/devmapper_wrapper_deferred_remove.go:func dmTaskDeferredRemoveFct(task *CDmTask) int` or `/pkg/devicemapper/devmapper_wrapper_deferred_remove.go:func dmTaskGetInfoWithDeferredFct(task *CDmTask, info *Info) int`.

samuelkarp commented:
@calavera, I've rebased again and now everything is passing.  What are the next steps here?

thaJeztah commented:
I think this only needs one more docs maintainer to review and LGTM

ping @SvenDowideit @moxiegirl for review :-)

moxiegirl commented:
@samuelkarp Woot there it is. LGTM

samuelkarp commented:
:grinning:

rodlogic commented:
Can this be used with docker 1.7.x? If so, how?

cpuguy83 commented:
@rodlogic Nope.

rodlogic commented:
@cpuguy83 Does that also mean that docker drivers are statically linked into the docker executable?

cpuguy83 commented:
@rodlogic All the drivers are built-ins and are part of the code-base... I think the journald driver makes some calls to some C libs, which can only be dynamically linked at compile time.

Golang does not currently support dynamically loaded modules/plugins.

pandrew commented:
If aws has one unified way to collect logs and we know that the driver relates to logging a more suitable name would be ```aws``` or ```cloudwatch```.

cpuguy83 commented:
@samuelkarp Can you shoot me an email, brian.goff@docker.com, looking to highlight cloudwatch support in Docker but need to get connect to the right folks on your team.

samuelkarp commented:
@cpuguy83 Will do.  Let me locate the right folks.
